app/services/scraper/bedelias.rb

```diff
@@ -0,0 +1,141 @@
+  rescue
+    Rails.logger.info save_screenshot
+    raise
+  end
+
+  private
+
+  def go_to_groups_and_subjects_page
+    visit "https://bedelias.udelar.edu.uy"
+    click_on "PLANES DE ESTUDIO"
+    click_on "Planes de estudio / Previas"
+
+    execute_script("$.fx.off = true;") # Disable jQuery effects
+
+    find('h3', text: 'TECNOLOGÍA Y CIENCIAS DE LA NATURALEZA').click
+    find('span', text: 'FING - FACULTAD DE INGENIERÍA', visible: false).click
+
+    within('tr', text: 'INGENIERIA EN COMPUTACION', match: :prefer_exact) do
+      find('.ui-row-toggler').click
+    end
+    within('.ui-expanded-row-content', text: 'Planes') do
+      find('tr', text: '1997').click_on "Ver más datos"
+    end
+  end
+
+  def process_groups_and_subjects(groups, subjects)
+    node_type = "[data-nodetype='Grupo']"
+    selector = ".arbolComposicion #{node_type}:not(:has(#{node_type})) > .ui-treenode-content > .ui-treenode-label"
+
+    all(selector, visible: false).each do |node|
+      group_code, name, credits = CODE_NAME_CREDITS_REGEX.match(node.text(:all)).captures
+      groups[group_code] = { code: group_code, name:, min_credits: credits.to_i }
+
+      node.all(:xpath, '..//..//li[@data-nodetype="Materia"]/span', visible: false).each do |subject_node|
+        code, name, credits = CODE_NAME_CREDITS_REGEX.match(subject_node.text(:all)).captures
+        subjects[code] = { code:, name:, credits: credits.to_i, has_exam: false, subject_group: group_code }
+      end
+    end
+  end
+
+  def go_to_prerequisites_page
+    click_on 'Sistema de previaturas'
+  end
+
+  def process_prerequisites(subjects)
+    total_pages = /\(\d+ de (\d+)\)/.match(find(".ui-paginator-current").text).captures[0]
+
+    1.upto(MAX_PAGES || total_pages.to_i).flat_map do |page|
+      Rails.logger.info "Page #{page}"
+
+      go_to_page(page, total_pages)
+
+      all('[data-ri]').map { |node| node['data-ri'] }.map do |row_id|
+        go_to_page(page, total_pages)
+
+        row = find("[data-ri='#{row_id}']")
+        name, type = row.all('td')[0..1]
+        Rails.logger.info "#{name.text} (#{type.text})"
+
+        subject_code = /\A(\w+) - .*\z/.match(name.text).captures[0]
+        is_exam = type.text == "Examen"
+
+        subjects[subject_code][:has_exam] ||= is_exam if subjects[subject_code]
+
+        row.click_on "Ver más"
+
+        prereq = Scraper::PrerequisitesTree.prerequisite_tree(find("[data-rowkey='root']"))
+
+        click_on 'Volver'
+
+        prereq.merge(subject_code:, is_exam:)
+      end
+    end
+  end
+
+  def go_to_page(page, total_pages)
+    if page > all('.ui-paginator-page').last.text.to_i
+      find('.ui-paginator-last').click
+      has_selector?(".ui-paginator-current", text: "(#{total_pages} de #{total_pages})")
+    end
+
+    find('.ui-paginator-page', text: page, match: :prefer_exact).click
+    has_selector?(".ui-paginator-current", text: "(#{page} de #{total_pages})")
+  end
```

Comment:
We're taking advantage of the fact that we only have `16` pages and if you stand on the last page you can see some of the page numbers that are also seen if you stand on the first page. That's not necessarily a bad thing, but it was hard to understand how was this working when I first looked at it.

IMHO I kinda prefer the previous implementation of this method which, even tho you would make one extra navigation when going to pages `15` and `16`, felt easier to understand at least for me 🙂

Reply:
yeah i think the previous approach was more correct, but i'm more worried about speed, because they might never add enough pages for this to be a problem, so we would be making unnecessary navigations. And also with this approach it will throw an error if there are more pages to make the requested page not visible.

Considering this i think i would prefer to keep it this way for the sake of speed, even though some time in the future this might throw an error. I added a comment explaining this in 3e77a49. Let me know if this makes sense or you still prefer the previous strategy